pkg/controllers/externalsecret/externalsecret_controller.go

```diff
@@ -128,345 +167,552 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
+		if apierrors.IsConflict(err) {
+			log.V(1).Info("conflict while updating secret, will requeue")
+			return ctrl.Result{Requeue: true}, nil
+		}
+
+		// detect errors indicating that we failed to set ourselves as the owner of the secret
+		// NOTE: this error cant be fixed by retrying so we don't return an error (which would requeue immediately)
+		if errors.Is(err, ErrSecretSetCtrlRef) {
+			r.markAsFailed(msgErrorBecomeOwner, err, externalSecret, syncCallsError.With(resourceLabels))
+			return ctrl.Result{}, nil
+		}
+
+		// detect errors indicating that the secret has another ExternalSecret as owner
+		// NOTE: this error cant be fixed by retrying so we don't return an error (which would requeue immediately)
+		if errors.Is(err, ErrSecretIsOwned) {
+			r.markAsFailed(msgErrorIsOwned, err, externalSecret, syncCallsError.With(resourceLabels))
+			return ctrl.Result{}, nil
+		}
+
+		// detect errors indicating that the secret is immutable
+		// NOTE: this error cant be fixed by retrying so we don't return an error (which would requeue immediately)
+		if errors.Is(err, ErrSecretImmutable) {
+			r.markAsFailed(msgErrorUpdateImmutable, err, externalSecret, syncCallsError.With(resourceLabels))
+			return ctrl.Result{}, nil
+		}
+
+		r.markAsFailed(msgErrorUpdateSecret, err, externalSecret, syncCallsError.With(resourceLabels))
 		return ctrl.Result{}, err
 	}
 
-	r.markAsDone(&externalSecret, start, log)
+	r.markAsDone(externalSecret, start, log, esv1beta1.ConditionReasonSecretSynced, msgSynced)
+	return r.getRequeueResult(externalSecret), nil
+}
+
+// getRequeueResult create a result with requeueAfter based on the ExternalSecret refresh interval.
+func (r *Reconciler) getRequeueResult(externalSecret *esv1beta1.ExternalSecret) ctrl.Result {
+	// default to the global requeue interval
+	// note, this will never be used because the CRD has a default value of 1 hour
+	refreshInterval := r.RequeueInterval
+	if externalSecret.Spec.RefreshInterval != nil {
+		refreshInterval = externalSecret.Spec.RefreshInterval.Duration
+	}
+
+	// if the refresh interval is <= 0, we should not requeue
+	if refreshInterval <= 0 {
+		return ctrl.Result{}
+	}
+
+	timeSinceLastRefresh := 0 * time.Second
+	if !externalSecret.Status.RefreshTime.IsZero() {
+		lastRefreshTime := externalSecret.Status.RefreshTime.Time
+		timeSinceLastRefresh = time.Since(lastRefreshTime)
+	}
+
+	// if the last refresh time is in the future, we should requeue immediately
+	// this should not happen as we call this function after updating the last refresh time
+	// to when the reconciliation started
+	if timeSinceLastRefresh < 0 {
+		return ctrl.Result{Requeue: true}
+	}
 
-	return ctrl.Result{
-		RequeueAfter: refreshInt,
-	}, nil
+	if timeSinceLastRefresh < refreshInterval {
+		// requeue after the remaining time
+		return ctrl.Result{RequeueAfter: refreshInterval - timeSinceLastRefresh}
+	} else {
+		// requeue immediately
+		return ctrl.Result{Requeue: true}
+	}
 }
```

Comment:
```suggestion
	}

	// requeue immediately
	return ctrl.Result{Requeue: true}
}
```

Reply:
At a certain point, I feel like you must have a vendetta against `else` statements!

Reply:
But I have removed it in https://github.com/external-secrets/external-secrets/pull/4086/commits/d40d76960900a1be45e13cb3110b06bc6cb9e84f

Reply:
I actually do. :D At least if there is a better way to return. Mostly because they significantly increase reading load and the score. :D That pesky score is above 15 anyways, but I came to accept that because the reconcile loop is significantly more complex.

At some point I want to refactor that, and I will. For now, I will accept this fate. :D